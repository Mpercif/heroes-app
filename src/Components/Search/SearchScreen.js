import React, { useMemo } from 'react'
import queryString from 'query-string'
import { useForm } from '../../Hooks/useForm';
import { HeroeCard } from '../Heroes/HeroeCard';
import { useLocation } from 'react-router';
import { getHeroesByName } from '../../Selectors/getHeroesByName';

export const SearchScreen = ({history}) => {

    const {search}  = useLocation()

    const {q = ''} = queryString.parse(search)

    const [{hero}, handleInputChange] = useForm({hero: q})

    const heroesFiltered =  useMemo(() =>getHeroesByName(q), [q])

    const handleSearch = (e)=>{
        e.preventDefault();
        history.push(`?q=${hero}`)
    }

    return (
        <div>
            <h1>SearchScreen</h1>
            <hr />

            <div className="row">

                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />

                    <form onSubmit={handleSearch}>
                        <input 
                            autoComplete='off'
                            type="text"
                            name="hero"
                            value={hero}
                            placeholder="Find your hero !"
                            className="form-control"
                            onChange={handleInputChange}
                        />

                        <button 
                            type="submit"
                            className="btn btn-outline-primary mt-2"
                        >
                            Search...
                        </button>

                    </form>

                </div>

                <div className="col-7">

                    <h4>Results...</h4>
                    <hr />

                    {
                        (q === '') 
                            ? 
                                <p className='alert alert-info'>Busca un heroe...</p> 
                            :
                                (heroesFiltered.length === 0) && 
                                    <p className='alert alert-info'>No se encontro el heroe con {q}</p>
                    }

                    {
                        heroesFiltered.map(hero =>(
                            <HeroeCard 
                                key={hero.id}
                                {...hero}
                            />
                        ))
                    }

                </div>

            </div>

        </div>
    )
}
