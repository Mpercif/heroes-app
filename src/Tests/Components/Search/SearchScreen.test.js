import { mount } from 'enzyme';
import React from 'react';
import { MemoryRouter, Route } from 'react-router';
import { SearchScreen } from '../../../Components/Search/SearchScreen';


describe('Pruebas en el componente <SearchScreen />', () => {

    const wrapper = mount(
        <MemoryRouter initialEntries={['/search']} >
            <Route 
                path='/search'
                component={SearchScreen}
            />
        </MemoryRouter>
    )
    
    test('Debe de mostrarse correctamente con valores por defecto ', () => {
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('.alert-info').text().trim()).toBe('Busca un heroe...')
    })  

    test('Debe de mostrar a batman y el input con el valor del queryString ', () => {
        
        const newWrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']} >
            <Route 
                path='/search'
                component={SearchScreen}
            />
        </MemoryRouter>
        )

        expect(newWrapper).toMatchSnapshot()
        expect(newWrapper.find('input').prop('value')).toBe('batman')
        expect(newWrapper.find('HeroeCard').exists()).toBe(true )
    })

    test('Debe de mostrar un error si no se encuntra en el Hero ', () => {
        
        const newWrapper2 = mount(
            <MemoryRouter initialEntries={['/search?q=batman312321']}>
                <Route 
                    path='/search'
                    component={SearchScreen}
                />
            </MemoryRouter>
        )

        expect(newWrapper2.find('.alert-info').text().trim()).
                                toBe('No se encontro el heroe con batman312321')

        expect(newWrapper2).toMatchSnapshot()
    })

    test('Debe llamar el push del history ', () => {
        
        const historyMock = {
            push: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman312321']} >
                <Route 
                    path='/search'
                    component={()=> <SearchScreen history={historyMock} /> }
                />
            </MemoryRouter>
        )

        wrapper.find('input').simulate('change',{
            target:{
                name: 'hero',
                value: 'batman'
            }
        })

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        })

        expect(historyMock.push).toHaveBeenCalledWith(`?q=batman`)
    })
    
    
    
})
