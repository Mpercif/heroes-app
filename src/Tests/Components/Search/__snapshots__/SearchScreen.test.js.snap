// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <SearchScreen /> Debe de mostrar a batman y el input con el valor del queryString  1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="hero"
          onChange={[Function]}
          placeholder="Find your hero !"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results...
      </h4>
      <hr />
      <div
        className="card mb-3"
        style={
          Object {
            "maxWidth": 350,
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="dc-batman"
              className="img-fluid rounded-start"
              src="./assets/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-tittle"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Detective Comics #27
              </p>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <a
                href="/heroe/dc-batman"
                onClick={[Function]}
              >
                Leer mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente <SearchScreen /> Debe de mostrar un error si no se encuntra en el Hero  1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="hero"
          onChange={[Function]}
          placeholder="Find your hero !"
          type="text"
          value="batman312321"
        />
        <button
          className="btn btn-outline-primary mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results...
      </h4>
      <hr />
      <p
        className="alert alert-info"
      >
        No se encontro el heroe con 
        batman312321
      </p>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente <SearchScreen /> Debe de mostrarse correctamente con valores por defecto  1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="hero"
          onChange={[Function]}
          placeholder="Find your hero !"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results...
      </h4>
      <hr />
      <p
        className="alert alert-info"
      >
        Busca un heroe...
      </p>
    </div>
  </div>
</div>
`;
